### PR TITLE
Web Worker

### PR DESCRIPTION
**Pull request by hobnob**
Generating a level is slow to load on older machines. While this PR doesn't necessarily fix the issue it achieves 2 things:

1) We can now animate items while the dungeon generates (giving feedback rather than a static screen)
2) Opens the door for further workers

I suspect the actual issue is something to do with the updates to the model or the changing of the scene, as the generation itself returns in just a few seconds. On my old Mac the game scene itself runs at 1 FPS and I suspect this has more to do with the slow load than the dungeon generation. That said this is still valuable work for the reasons stated above

### CODE REVIEW

File: src/main/scala/roguelike/subsystems/WorkerSubSystem.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import indigo.shared.subsystems.SubSystemId
+import org.scalajs.dom
+import org.scalajs.dom.Event
+import roguelike.workers.IndigoWorker
+
+import scala.annotation.nowarn.apply
+import scala.collection.mutable
+import scala.scalajs.js
+
+final case class WorkerSubSystem[A <: js.Any, B <: js.Any](
```
Comment (davesmith00000) [inline src/main/scala/roguelike/subsystems/WorkerSubSystem.scala:20]: This type looks a bit suspicious `A <: js.Any` - isn't that just anything?

File: .scalafix.conf [hunk 2]
```diff
@@ -13,7 +13,7 @@ DisableSyntax {
   noNulls = true
   noReturns = true
   noWhileLoops = true
-  noAsInstanceOf = true
```
[the remark this answers is not in the document]
Comment (davesmith00000) [reply .scalafix.conf:16]: You could just put it above `IndigoWorker` for now?

File: assets/dungeon-gen-worker.js [hunk 3]
```diff
@@ -0,0 +1,2 @@
+importScripts("../scripts/roguelike-fastopt.js");
+DungeonGenWorker.doWork();
```
[the remark this answers is not in the document]
Comment (hobnob) [reply assets/dungeon-gen-worker.js:2]: VS wasn't configured properly :-/

File: src/main/scala/roguelike/GeneratingLevelScene.scala [hunk 4]
```diff
@@ -26,16 +34,29 @@ object GeneratingLevelScene extends Scene[Size, Model, ViewModel]:
...
   def updateModel(
       context: FrameContext[Size],
       model: Model
   ): GlobalEvent => Outcome[Model] =
     case GenerateLevel =>
+      Outcome(model).addGlobalEvents(
+        workerSubSystem.WorkerEvent.Send(new DungeonGenConfig {
+          val seed: Double      = context.dice.seed.toDouble
+          val currentLevel: Int = model.currentFloor
+        })
```
[the remark this answers is not in the document]
Comment (hobnob) [reply src/main/scala/roguelike/GeneratingLevelScene.scala:51]: Oh cool - I didn't think realise you could do that :-)

File: src/main/scala/roguelike/model/Model.scala [hunk 5]
```diff
@@ -313,10 +313,10 @@ final case class Model( // TODO: Should there be a GameModel class too? (Similar
         val p =
           gameMap
             .getPathTo(player.position, target, Batch.empty)
-            .tail
+            .drop(1)
         if p.nonEmpty then
           val next      = p.head
-          val remaining = p.tail
+          val remaining = p.drop(1)
```
[the remark this answers is not in the document]
Comment (hobnob) [reply src/main/scala/roguelike/model/Model.scala:319]: Ah I see - yeah I didn't know where the exception was coming from, so I just got rid of them all :-P

File: src/main/scala/roguelike/subsystems/WorkerSubSystem.scala (hunk 1 above)
Comment (hobnob) [reply src/main/scala/roguelike/subsystems/WorkerSubSystem.scala:20]: It's anything that can be converted to JS - this is so you can pass lists, scalars, objects etc. A bit messy, but the whole worker ecosystem is messy :-/

File: src/main/scala/roguelike/model/Model.scala [hunk 6]
```diff
@@ -387,7 +387,8 @@ final case class Model( // TODO: Should there be a GameModel class too? (Similar
       )
 
     case GameEvent.PlayerDescended =>
-      Model.genNextFloor(context.dice, this)
+      Outcome(this)
+    // Model.genNextFloor(context.dice, this)
```
[the remark this answers is not in the document]
Comment (hobnob) [reply src/main/scala/roguelike/model/Model.scala]: All fixed :-)

**Pull request merged**